Jl/caip multichain/handle accounts provider authorize

jiexi opened the pull request:
NOTE: We may want to get [this child branch](https://github.com/MetaMask/metamask-extension/pull/25713) merged into this branch first before making the remaining changes to accounts behavior


<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

* Checks if accounts specified in the `accounts` property exist in the Keyring API
* Requests the `eth_accounts` permission using those accounts (makes the assumption they are evm)
* Removes random property sessions from my initial implementation

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25708?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jiexi committed:
initial scope spec isValidScope

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
Add isValidScope spec

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
spec isSupportedNotification. spec and update isSupportedScopeString

jiexi committed:
spec flattenScope

jiexi committed:
fix and spec mergeScopeObject

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
throw if scopes if defined in mergeFlattenedScopes

jiexi committed:
lint

jiexi committed:
fix ts lint

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
Remove randomSessionProperties

jiexi committed:
request accounts from provider_authorize

jiexi committed:
handle accounts. cleanup

jiexi commented on app/scripts/lib/multichain-api/scope.ts:
```diff
@@ -54,15 +58,32 @@ export type Caip25Authorization =
       sessionProperties?: Record<string, unknown>;
     });
 
+export const parseScopeString = (
```
spec this

jiexi commented on app/scripts/lib/multichain-api/scope.ts:
```diff
@@ -144,7 +177,7 @@ enum KnownCaipNamespace {
 export const isSupportedScopeString = (
   scopeString: string,
-  findNetworkClientIdByChainId?: (chainId: Hex) => NetworkClientId,
+  findNetworkClientIdByChainId: (chainId: Hex) => NetworkClientId,
```
update this spec

jiexi commented on app/scripts/lib/multichain-api/scope.ts:
```diff
@@ -184,6 +215,31 @@ export const isSupportedScopeString = (
   return false;
 };
 
+export const isSupportedAccount = (
```
spec this

jiexi replied:
i don't think this method needs to exist anymore

jiexi committed:
add isSupportedAccount spec

jiexi committed:
lint

jiexi committed:
Remove top of file caip spec comments

jiexi committed:
lint

shanejonas commented on app/scripts/lib/multichain-api/provider-authorize.js:
```diff
@@ -161,14 +143,29 @@ export const assertScopesSupported = (scopes, findNetworkClientIdByChainId) => {
+        // The below is made up
+        throw new EthereumRpcError(
+          5103,
+          'Requested accounts are not supported',
```
this introduces a way for a dapp to check if you have an account without popping up a UI

jiexi replied:
Good point. The only way to not leak this would be to always open the confirmation and only give the reject option to the user. The other way around this is to simply not honor the requested accounts. Which do you think is preferable?

adonesky1 replied:
I would think that we don't actually want to do validation against the accounts and only honor them (pre-select them in the connection UI) if they're valid. I don't think it needs to block the request. Say a user deletes that accoount in their wallet after having previously used it with a dapp the dapp might request it but then we would be failing the whole connection request when this validation failed.

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/scope-helper-specs

jiexi committed:
Merge branch 'jl/caip-multichain/scope-helper-specs' into jl/caip-multichain/handle-accounts-provider-authorize

jiexi commented on app/scripts/lib/multichain-api/scope.ts:
```diff
@@ -184,6 +195,31 @@ export const isSupportedScopeString = (
+      try {
+        return getInternalAccounts().some(
+          (internalAccount) =>
+            ['eip155:eoa', 'eip155:erc4337'].includes(internalAccount.type) &&
```
https://github.com/MetaMask/keyring-api/blob/f4b34112c82dda0f159769e9f37d08c3c0bdbf96/src/api/account.ts#L11-L14

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/handle-accounts-provider-authorize

adonesky1 commented on app/scripts/lib/multichain-api/provider-authorize.js:
```diff
@@ -189,8 +192,19 @@ export const processScopes = (
 export async function providerAuthorizeHandler(req, res, _next, end, hooks) {
-  const { requiredScopes, optionalScopes, sessionProperties, ...restParams } =
-    req.params;
+  // TODO: Does this handler need a rate limiter/lock like the one in eth_requestAccounts?
```
yes I would think so

adonesky1 commented on app/scripts/lib/multichain-api/provider-authorize.js:
```diff
@@ -93,7 +74,10 @@ export const flattenScopes = (scopes) => {
+  scopes,
+  { findNetworkClientIdByChainId, getInternalAccounts },
+) => {
   // TODO: Should we be less strict validating optional scopes? As in we can
   // drop parts or the entire optional scope when we hit something invalid which
   // is not true for the required scopes.
```
this is a good question 🤔 for now lets be strict on this but we may want to change that later

adonesky1 commented on app/scripts/lib/multichain-api/provider-authorize.js:
```diff
@@ -220,11 +227,44 @@ export async function providerAuthorizeHandler(req, res, _next, end, hooks) {
+      );
+    });
+
+    if (accounts.length > 0) {
+      await hooks.requestPermissions(
+        { origin },
+        {
+          [RestrictedMethods.eth_accounts]: {
+            caveats: [
+              {
+                type: CaveatTypes.restrictReturnedAccounts,
+                value: unique(accounts),
+              },
+            ],
+          },
+        },
+      );
+    }
```
is the idea here that we're just doing this for testing until we add the multichain permission middleware?

jiexi commented on app/scripts/lib/multichain-api/scope.ts:
```diff
@@ -184,6 +195,31 @@ export const isSupportedScopeString = (
 
+export const isSupportedAccount = (
+  account: CaipAccountId,
+  getInternalAccounts: () => InternalAccount[],
```
should not do this anymore

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/handle-accounts-provider-authorize

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/handle-accounts-provider-authorize

jiexi committed:
Merge branch 'caip-multichain' into jl/caip-multichain/handle-accounts-provider-authorize

jiexi committed:
fix spec

jiexi committed:
Jl/caip multichain/provider authorize specs (#25713)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

* Move provider_authorize helpers into scope utils file
* Reorganize scope utils for easier testing
* Spec new helpers in scope utils
* Add provider_authorize spec

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25713?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

shanejonas committed:
fix: remove accounts validation and get the accounts from requestPerm… (#25773)

…issions popup for now


<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25773?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

shanejonas committed:
Merge branch 'caip-multichain' into jl/caip-multichain/handle-accounts-provider-authorize

adonesky1 commented on app/scripts/lib/multichain-api/scope/assert.ts:
```diff
@@ -0,0 +1,103 @@
+    );
+  }
+
+  if (accounts) {
+    const accountsSupported = accounts.every((account) =>
+      isSupportedAccount(account, getInternalAccounts),
+    );
+
+    if (!accountsSupported) {
+      // TODO: There is no error code or message specified in the CAIP-25 spec for when accounts are not supported
+      // The below is made up
+      throw new EthereumRpcError(5103, 'Requested accounts are not supported');
```
weren't we going to get rid of this?

shanejonas replied:
i commented out the validation part so it doesnt get called. also a little bit weird to throw a jsonrpc specific error from this helper since it could get used in validation for the caveats
